hydroshare_jupyter_sync/lib/filesystem/aggregate_fs_map.py

```diff
@@ -0,0 +1,104 @@
+        self._map_fn(lambda o: o.add_resource(resource_id))
+
+    def delete_resource(self, resource_id: ResourceId) -> None:
+        """Remove resource from local and remote FSMap instances"""
+        self._map_fn(lambda o: o.delete_resource(resource_id))
+
+    def update_resource(self, resource_id: ResourceId) -> None:
+        """Update a local and remote resource"""
+        self._map_fn(lambda o: o.update_resource(resource_id))
+
+    # IEntityFSMap implementations
+
+    def add_resource_file(
+        self, resource_id: ResourceId, relative_resource_file: Union[Path, str]
+    ) -> None:
+        """Add resource file to local map. Does not interact with the remote FS map."""
+        self.local_map.add_resource_file(resource_id, relative_resource_file)
+
+    def delete_resource_file(
+        self, resource_id: ResourceId, relative_resource_file: Union[Path, str]
+    ) -> None:
+        """Remove resource file from local map. Does not interact with the remote FS map."""
+        self.local_map.delete_resource_file(resource_id, relative_resource_file)
+
+    def update_resource_file(
+        self, resource_id: ResourceId, relative_resource_file: Union[Path, str]
+    ):
+        """Update resource file in local map. Does not interact with the remote FS map."""
+        self.local_map.update_resource_file(resource_id, relative_resource_file)
+
+    # other methods
+
+    def get_sync_state(self) -> AggregateFSResourceMapSyncStateCollection:
+        """Get sync status of all resources in local and remote maps."""
+        return AggregateFSResourceMapSyncStateCollection.from_aggregate_map(
+            aggregate_fs_map=self
+        )
+
+    def get_resource_sync_state(
+        self, resource_id: ResourceId
+    ) -> AggregateFSResourceMapSyncState:
+        """Get sync status for a given resource between the local and remote map."""
+        if resource_id in self.local_map and resource_id in self.remote_map:
+            local_resource = self.local_map[resource_id]
+            remote_resource = self.remote_map[resource_id]
+            return AggregateFSResourceMapSyncState.from_resource_maps(
+                local_resource_map=local_resource, remote_resource_map=remote_resource
+            )
+
+        error_message = (
+            f"ResourceID: {resource_id}, does not exist in local_map or remote_map.\n"
+            f"{self.local_map.keys()=}\n"
+            f"{self.remote_map.keys()=}"
+        )
+        raise AggregateFSMapResourceMembershipError(error_message)
+
+    # helper methods
+    def _map_fn(self, fn, *args, **kwargs) -> Tuple[T]:
+        """Map a function call over the LocalFSMap and RemoteFSMap. Results returned in that order."""
+        res = []
+        for o in [self.local_map, self.remote_map]:
+            # append result
+            res.append(fn(o, *args, **kwargs))
+
+        # cast to tuple for in adherence with standard pythonic practices
+        return tuple(res)
```

Comment:
These results aren't returned from the public methods.  I do not see any benefit to this helper method outside of guaranteeing the result order.  Most of the methods in this class are aware of the local/remote maps... I think it'd be easier to read if you just called the methods on the maps without the helper function.  It could be useful if the number maps were to increase.

Reply:
> These results aren't returned from the public methods.

Right, but they could be. I wrote it to be a little more general than the current use cases for future features that could be added.

Yeah, I don't disagree with you -- I was mainly trying to cut down on code duplication.  I wrote it this way with the idea that another map could be added in the future (like you mentioned). I think I will leave it as is. I think a good compromise might be to make it a public method on the class so if others wanted to extend the functionality, there is a way to do that easily. What are your thoughts?